File: addondev/plugins/org.addondev/src/org/addondev/parser/javascript/Lexer.java
package org.addondev.parser.javascript;


import java.util.HashMap;
import java.util.Map;

public class Lexer {
	private int tok;
	 private String val;
	 public LexerReader reader;
	 private int offset;

	private static Map<String, Integer> reserved = new HashMap<String, Integer>(); // 予約語を保持する

	static { // 予約語を登録

		//reserved.put("if", new Integer(TokenType.IF));
		//reserved.put("else", new Integer(TokenType.ELSE));
		//reserved.put("while", new Integer(TokenType.WHILE)); // while
		reserved.put("function", new Integer(TokenType.FUNCTION));
		reserved.put("var", new Integer(TokenType.VAR));
		reserved.put("const", new Integer(TokenType.CONST));
		reserved.put("new", new Integer(TokenType.NEW));
		//reserved.put("true", new Integer(TokenType.TRUE));
		//reserved.put("false", new Integer(TokenType.FALSE));
		
		reserved.put("try", new Integer(TokenType.TRY));
		reserved.put("catch", new Integer(TokenType.CATCH));
		reserved.put("finally", new Integer(TokenType.FINALLY));
		
		reserved.put("let", new Integer(TokenType.LET));
	}

	public Lexer(String src) {
		reader = new LexerReader(src); // readerにはLexerReaderをセットする！
	}

	public String value() {
		return val;
	}

	public boolean advance() {
		skipWhiteSpace();
		int c = reader.read();
		if (c < 0) {
			return false;
		}
		switch (c) {
		case '{':
		case '}':
		case ':':
		case ';':
		case ',':
	    case '(':
	    case ')':
	    case '.':	    	
	    case '^':
	    case '~':
	    case '[':
	    case ']':
	    case '\\':
	    case '?':
	    //case '$':
	    case '#':
	    case '!':
	    case '*':
	    case '>':
	    case '<':
	    case '+':
	    case '-':
	    case '|':
	    case '&':
	    case '%':
	    case '@':
			tok = c;
			break;
		case '=':
	        c = reader.read();
	        if(c == '='){
	          tok = TokenType.EQ;    // '=='
	        }else{
	          reader.unread(c);
	          tok = '=';             // '='
	        }			
			break;
		case '/':
	        c = reader.read();   // 次の文字が
	        if(c == '/'){        // '/'だったら
	          skipLineComment(); //   １行コメントとして読み飛ばし
	          return advance();  //   次のトークンを読みにいく。
	        }else if(c == '*'){  // '*'だったら
	        	c = reader.read(); // /** だったら
	        	if(c == '*')
	        	{
	        		tok = TokenType.JSDOC;
	        		jsDoc();
	        		//return advance();  //   次のトークンを読みにいく。
	        	}
	        	else
	        	{
		          skipComment();     //   複数行コメントとして読み飛ばし
		          return advance();  //   次のトークンを読みにいく。
	        	}
	        }else{               // それ以外なら
	          reader.unread(c);  // １文字戻しておいて
	          tok = '/';         // 普通に演算子'/'として処理する。
	        }
			break;
		case '"':
		case '\'':
			//buf = new StringBuffer();
			lexString((char)c);
			tok = TokenType.STRING;
			break;
		default:
			if(Character.isDigit((char)c)){
		          reader.unread(c);
		          lexDigit();
		          tok = TokenType.INT;				 
			}
			else
			{
				reader.unread(c);
				lexSymbol(); 
			}
		}
		return true;
	}
	
	  private void lexDigit() {
		    int num = 0;
		    while(true){
		      int c = reader.read();
		      if(c < 0){
		        break;
		      }
		      if(!Character.isDigit((char)c)){
		        reader.unread(c);         // returnする前にunread()する！
		        break;
		      }
		      num = (num * 10) + (c - '0');
		    }
		    val = String.valueOf(num);
		  }	
	
	//private StringBuffer buf = new StringBuffer();
	private void lexString(char ch) {
		//StringBuffer buf = new StringBuffer();
		StringBuilder buf = new StringBuilder();
		// TODO Auto-generated method stub
		while (true) {
			int c = reader.read();
			if (c < 0) {
				//throw new Exception("文字列中でファイルの終わりに到達しました。");
				break;
			}
			//if (c == '"') {
			if (c == ch) {
				break;
			} else if (c == '\\') { // (A)
				c = reader.read();
				if (c < 0) {
					//throw new Exception("文字列中でファイルの終わりに到達しました。");
					break;
				}
			}
			buf.append((char) c);
		}
		val = buf.toString();	
	}

	private void skipComment() {
		// TODO Auto-generated method stub
	    int c = '\0';
	    while(true){
	      c = reader.read();
	      if(c < 0){
	        //throw new Exception("コメント中にファイルの終端に到達しました。");
	    	  break;
	      }
	      if(c == '*'){
	        c = reader.read();
	        if(c == '/'){
	          break;
	        }
	      }
	    }		
	}

	private void skipLineComment() {
		// TODO Auto-generated method stub
	    int c;
	    while((c = reader.read()) != '\n'){
	      if(c < 0){
	        //throw new Exception("コメント中にファイルの終端に到達しました。");
	    	  break;
	      }
	    }
	    reader.unread(c);		
	}

	private void skipWhiteSpace() {
		// TODO Auto-generated method stub
		int c = reader.read();
		char ch = (char)c;
		//while ((c != -1) && Character.isWhitespace((char) c)) {
		while ((c != -1) && (Character.isWhitespace(ch) || (ch == '\t') || (ch == '\n') || (ch == '\r'))) {
			c = reader.read();
			ch = (char)c;
		}
		reader.unread(c);
	}

	private void lexSymbol() {
		offset = reader.offset();
		
		tok = TokenType.SYMBOL;
		StringBuilder buf = new StringBuilder();
		while (true) {
			int c = reader.read();
			if (c < 0) {
				// throw new Exception("ファイルの終わりに到達しました。");
			}
			if (!Character.isJavaIdentifierPart((char) c)) {
				reader.unread(c);
				break;
			}
			buf.append((char) c);
		}
		String s = buf.toString();
		val = s;

		if (reserved.containsKey(s)) { // (A)
			tok = ((Integer) reserved.get(s)).intValue();
		}
	}
	
	private void jsDoc() {
		// TODO Auto-generated method stub
		StringBuilder buf = new StringBuilder();
		buf.append("/**");
	    int c = '\0';
	    while(true){
	      c = reader.read();
	      if(c < 0){
	        //throw new Exception("コメント中にファイルの終端に到達しました。");
	    	  break;
	      }
	      buf.append((char) c);
	      if(c == '*'){
	        c = reader.read();
	        if(c == '/'){
	        	buf.append("/");
	    		String s = buf.toString();
	    		val = s;
	          break;
	        }
	      }
	    }		
	}
	
	public void jsRex() {
		// TODO Auto-generated method stub
		int d = 0;
	    int c = '\0';
	    while(true){
	    	d = c;
	      c = reader.read();
	      if(c < 0){
	        //throw new Exception("コメント中にファイルの終端に到達しました。");
	    	  break;
	      }
	      
	      //if(c != '\\'){
	        //c = reader.read();
	        if(d != '\\' && c == '/'){
	        //if(c == '/'){
	        	
	          break;
	        }
	      //}
	    }	
	}

	public int token() {
		// TODO Auto-generated method stub
		 return tok;
	}
	
	public int offset()
	{
		offset = reader.offset();
		return offset;
	}
	
	public void un(int c)
	{
		reader.unread(c);
	}
}

File: test/gcsTextEdit/gcsTextEdit/Parser/Lexer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using YYS;
using YYS.Parser;

namespace YYS.Parser {

    public class Lexer {

        //public string Value { get; set; }

        public Tuple<int, int, Rule> OffsetLenAttr;

        private TokenType tok;
        public LexerReader reader;

        //private Rule resultRule;

        public bool isNextLine = false;

        private Dictionary<String, Rule> ruleDic = new Dictionary<String, Rule>();

        private Dictionary<String, MultiLineRule> multiRuleDic = new Dictionary<String, MultiLineRule>();
        private Dictionary<String, MultiLineRule> multiRuleEndDic = new Dictionary<String, MultiLineRule>();

        private List<KeywordRule> keyWordRules = new List<KeywordRule>();

        public void ClearRule() {
            ruleDic.Clear();
            multiRuleDic.Clear();
            multiRuleEndDic.Clear();
            keyWordRules.Clear();
        }

        public void AddRule(List<Rule> rules) {
            foreach (var item in rules) {
                this.AddRule(item);
            }
        }
        public void AddRule(Rule rule) {
            if (rule is KeywordRule) {
                keyWordRules.Add(rule as KeywordRule);
            }
            else {
                if (rule is MultiLineRule) {
                    multiRuleDic.Add(((MultiLineRule)rule).start, (MultiLineRule)rule);
                    multiRuleEndDic.Add(((MultiLineRule)rule).end, (MultiLineRule)rule);
                }

                {
                    ruleDic.Add(rule.start, rule);
                }
            }
        }

        public IText Src {
            get { return reader.Src; }
            set {
                if (reader == null) {
                    reader = new LexerReader(value);
                }
                else {
                    reader.Src = value;
                    tok = TokenType.TXT;
                    isNextLine = false;
                    //Value = null;
                }
            }
        }

        public TokenType token {
            get { return tok; }
        }

        public Lexer(IText src) {
            reader = new LexerReader(src);
        }

        public Lexer(LexerReader reader) {
            this.reader = reader;
        }

        public Lexer() {
            reader = new LexerReader();
        }

        public int Offset {
            get {
                return reader.offset();
            }
        }

        private int start;
        private int end;
        public void SetRange(int start, int end) {
            this.start = start;
            this.end = end;

            reader.setoffset(start);
        }

        public bool advance(Block preblock, Block curblock) {
            tok = TokenType.TXT;

            if (Src.Length == 0 && curblock.isLineHeadCmt == 1) {
                curblock.mRule = preblock.mRule;
                tok = TokenType.MultiLineAllLine;
                isNextLine = true;
                return true;
            }

            int c = reader.read();
            if (c == -1) {
                tok = TokenType.EOS;
                return false;
            }

            switch (c) {

                case ' ':
                case 0x3000:
                case '\t':
                    break;

                default:
                    if (curblock.isLineHeadCmt == 0) { //0: 行頭がブロックコメントの内部ではない
                        if (Char.IsDigit((char)c)) {
                            reader.unread();
                            lexDigit();
                            break;
                        }
                        else if (Util.isIdentifierPart((char)c)) {
                            reader.unread();
                            lexKeyWord();
                            break;
                        }
                        else {
                            reader.unread();
                            lexSymbol(curblock);
                            break;
                        }
                    }
                    else { //1: 行頭がブロックコメントの内部

                        if (Offset - 1 == 0) {
                            int index;
                            if ((index = Src.IndexOf(preblock.mRule.end, 0)) >= 0) {
                                string end = preblock.mRule.end;
                                var rule = multiRuleEndDic[end];
                                if (rule.Detected(end, reader)) {
                                    curblock.mRule = rule;
                                    tok = TokenType.MultiLineEnd;
                                    int len = rule.getLen(end, reader);
                                    reader.setoffset(len);
                                    OffsetLenAttr = new Tuple<int, int, Rule>(0, len, rule);
                                    return true;
                                }

                            }
                            else if(multiRuleDic.ContainsKey(preblock.mRule.start)){
                                var rule = multiRuleDic[preblock.mRule.start];
                                curblock.mRule = preblock.mRule;
                                tok = TokenType.MultiLineAllLine;
                                OffsetLenAttr = new Tuple<int, int, Rule>(0, Src.Length, rule);
                                reader.setoffset(Src.Length);
                                return true;
                            }

                            //{
                            //    reader.unread();
                            //    StringBuilder buf = new StringBuilder();
                            //    while (true) {
                            //        c = reader.read();
                            //        if (c == -1) {
                            //            break;
                            //        }
                            //        buf.Append((char)c);

                            //        string s = buf.ToString();
                            //        string end = preblock.mRule.end;

                            //        if (s.EndsWith(end)) {
                            //            if (multiRuleEndDic.ContainsKey(end)) {
                            //                var rule = multiRuleEndDic[end];
                            //                if (rule.Detected(end, reader)) {
                            //                    curblock.mRule = rule;
                            //                    //isNextLine = false;
                            //                    //tok = rule.token;
                            //                    tok = TokenType.MultiLineEnd;
                            //                    int len = rule.getLen(end, reader);
                            //                    reader.setoffset(len);
                            //                    OffsetLenAttr = new Tuple<int, int, Rule>(0, len, rule);
                            //                    //break;
                            //                    return true;
                            //                }
                            //            }
                            //        }
                            //    }

                            //    if (c == -1 && preblock.mRule != null && multiRuleDic.ContainsKey(preblock.mRule.start)) {
                            //        var enelem = multiRuleDic[preblock.mRule.start];
                            //        curblock.mRule = preblock.mRule;
                            //        //isNextLine = true;
                            //        //tok = TokenType.MultiLineStart;
                            //        tok = TokenType.MultiLineAllLine;
                            //        OffsetLenAttr = new Tuple<int, int, Rule>(0, Src.Length, enelem);
                            //        reader.setoffset(Src.Length);
                            //    }
                            //    break;
                            //}
                        }
                        else {
                            if (Char.IsDigit((char)c)) {
                                reader.unread();
                                lexDigit();
                            }
                            else if (Util.isIdentifierPart((char)c)) {
                                reader.unread();
                                lexKeyWord();
                            }
                            else {
                                reader.unread();
                                lexSymbol(curblock);
                            }
                        }
                    }
                    break;
            }
            return true;
        }

        private void skipWhiteSpace() {
            int c = reader.read();
            char ch = (char)c;

            while ((c != -1) && (Char.IsWhiteSpace(ch) || (ch == '\t') || (ch == '\n') || (ch == '\r'))) {
                c = reader.read();
                ch = (char)c;
            }
            reader.unread();
        }

        private void lexDigit() {
            int num = 0;
            while (true) {
                int c = reader.read();
                if (c < 0) {
                    break;
                }
                if (!Char.IsDigit((char)c)) {
                    reader.unread();
                    break;
                }
                num = (num * 10) + (c - '0');
            }
            tok = TokenType.Number;
        }

        private void lexKeyWord() {
            int offset = reader.offset() - 1;

            StringBuilder buf = new StringBuilder();
            while (true) {
                int c = reader.read();
                if (c < 0) {
                    break;
                }
                if (!Util.isIdentifierPart((char)c)) {
                    reader.unread();
                    break;
                }
                buf.Append((char)c);
            }
            String s = buf.ToString();

            foreach (var rule in keyWordRules) {
                if (rule.Detected(s, reader)) {
                    tok = rule.token;
                    int len = rule.getLen(s, reader);
                    reader.setoffset(offset + len);
                    OffsetLenAttr = new Tuple<int, int, Rule>(offset, len, rule);
                    break;
                }
            }
        }

        private void lexSymbol(Block curblock) {
            StringBuilder buf = new StringBuilder();
            int offset = reader.offset() - 1;

            while (true) {
                int c = reader.read();
                if (c == -1) {
                    return;
                }
                buf.Append((char)c);

                string s = buf.ToString();
                if (ruleDic.ContainsKey(s)) {
                    var rule = ruleDic[s];
                    if (rule.Detected(s, reader)) {
                        tok = rule.token;
                        int len = rule.getLen(s, reader);
                        OffsetLenAttr = new Tuple<int, int, Rule>(offset, len, rule);
                        reader.setoffset(offset + len);

                        if (rule is MultiLineRule) {
                            if (curblock != null) {
                                curblock.mRule = rule as MultiLineRule;
                                tok = TokenType.MultiLineStart;
                                //isNextLine = true;
                            }
                        }
                        break;
                    }
                }
                else if (multiRuleEndDic.ContainsKey(s)) {
                    var rule = multiRuleEndDic[s];
                    if (rule.Detected(s, reader)) {
                        //tok = rule.token;
                        tok = TokenType.MultiLineEnd;
                        int len = rule.getLen(s, reader);
                        OffsetLenAttr = new Tuple<int, int, Rule>(offset, len, rule);
                        reader.setoffset(offset + len);

                        if (curblock != null) {
                            curblock.mRule = rule as MultiLineRule;
                        }

                        break;
                    }
                }
            }
        }
    }
}
